### PR TITLE
Update to 0.3.0 of arkworks

### PR DESCRIPTION
**Pull request by mstraka100**
Updates the repository to use version `0.3.0` of `arkworks`. This is needed to stay up-to-date on the latest functionality and bug fixes contained in this dependency. Also updates `dkg-cli` to use the latest version of `ethers-rs`, necessary because the previously used version does not compile using the latest stable Rust compiler.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -20,3 +20,4 @@ react-native
 
 solidity/cache
 solidity/build
+*.swp
```
Comment (piersy) [inline .gitignore:23]: Youve added a couple of `.DS_Store` files in this PR
```suggestion
*.swp
.DS_Store
```

File: crates/dkg-cli/Cargo.toml [hunk 2]
```diff
@@ -1,15 +1,21 @@
 [package]
 name = "dkg-cli"
-version = "0.1.0"
-authors = ["Georgios Konstantopoulos <me@gakonst.com>"]
-edition = "2018"
+version = "0.3.0"
+authors = ["Georgios Konstantopoulos <me@gakonst.com>", "Michael Straka <mstraka100@gmail.com>"]
+edition = "2022"
```
Comment (palango) [inline crates/dkg-cli/Cargo.toml:5]: As far as I'm aware there's no 2022 edition: https://doc.rust-lang.org/edition-guide/index.html

File: crates/dkg-cli/Cargo.toml [hunk 3]
```diff
@@ -19,11 +25,15 @@ hex = "0.4.2"
 anyhow = "1.0.31"
 
 ethers = { git = "https://github.com/gakonst/ethers-rs", features = ["celo"] }
+ethers-solc = { git = "https://github.com/gakonst/ethers-rs" }
```
Comment (palango) [inline crates/dkg-cli/Cargo.toml:28]: Can you pin all git dependencies to commits so the build won't break in the future? See https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories

File: Cargo.toml [hunk 4]
```diff
@@ -2,7 +2,6 @@
 
 members = [ 
     "crates/dkg-core", 
-    "crates/dkg-cli",
```
Comment (palango) [inline Cargo.toml:5]: Why did you remove this?

File: crates/dkg-cli/src/actions.rs [hunk 5]
```diff
@@ -333,11 +344,13 @@ async fn wait_for_phase<P: JsonRpcClient, S: Signer>(
...
 
-fn parse_bundle<D: serde::de::DeserializeOwned>(bundle: &[Vec<u8>]) -> Result<Vec<D>> {
+fn parse_bundle<D: serde::de::DeserializeOwned>(
+    bundle: &[ethers::prelude::Bytes],
+) -> Result<Vec<D>> {
     bundle
         .iter()
-        .filter(|item| !item.is_empty()) // filter out empty items
-        .map(|item| Ok(bincode::deserialize::<D>(&item)?))
+        .filter(|item| !item.to_vec().is_empty()) // filter out empty items
```
Comment (palango) [inline crates/dkg-cli/src/actions.rs:352]: I cannot get the build to work, but maybe you can replace `!item.to_vec().is_empty()` with `!item.is_empty()`, which would reduce allocations.

File: crates/dkg-cli/src/actions.rs [hunk 6]
```diff
@@ -333,11 +344,13 @@ async fn wait_for_phase<P: JsonRpcClient, S: Signer>(
...
+        .map(|item| Ok(bincode::deserialize::<D>(&item.to_vec()[..])?))
```
Comment (palango) [inline crates/dkg-cli/src/actions.rs:353]: Same here, instead of going through a new `Vec`, you might be able to use `&item.as_slice()`

File: crates/dkg-cli/src/actions.rs [hunk 7]
```diff
@@ -70,50 +75,48 @@ pub async fn deploy(opts: DeployOpts) -> Result<()> {
...
+    M: Middleware,
     R: RngCore,
 {
     let provider = Provider::<Http>::try_from(opts.node_url.as_str())?;
-    let client = Arc::new(opts.private_key.parse::<Wallet>()?.connect(provider));
+    let wallet = opts.private_key.parse::<LocalWallet>()?;
+    let client = SignerMiddleware::new(provider, wallet);
+    let client = Arc::new(client);
 
+    //ethers::signers::Wallet, ethers::providers::Provider
```
Comment (palango) [inline crates/dkg-cli/src/actions.rs:119]: Maybe remove this

File: crates/dkg-cli/Cargo.toml [hunk 8]
```diff
@@ -19,11 +25,15 @@ hex = "0.4.2"
...
 futures = "0.3.5"
 async-trait = "0.1.35"
-tokio = { version = "0.2.21", features = ["macros"] }
+tokio = { version = "0.2.21", features = ["macros", "time"] }
```
Comment (palango) [inline crates/dkg-cli/Cargo.toml:31]: Is the version correct here, seems like ethers needs tokio 1.0+

File: crates/threshold-bls/src/sig/sig.rs [hunk 9]
```diff
@@ -33,7 +33,7 @@ pub trait Scheme: Debug {
     }
 }
 
-/// SignatureScheme is the trait that defines the operations of a sinature
+/*/// SignatureScheme is the trait that defines the operations of a sinature
```
Comment (palango) [inline crates/threshold-bls/src/sig/sig.rs:36]: Do these extra comments have some meaning?

File: crates/dkg-cli/README.md [hunk 10]
```diff
@@ -1,143 +0,0 @@
-# Distributed Key Generation CLI
```
Comment (palango) [inline crates/dkg-cli/README.md:1]: Is this outdated or why did you remove this?

**Review by palango — COMMENTED** [review]
This is a massive amount of changes, crazy that you managed to do all this all at once.

I left some comments on stuff that I notices, but all the crypto stuff is over my head.

For some reason I didn't manage to go through the build-step, so some comments might be wrong.
```
➜  celo-threshold-bls-rs git:(straka/arkworks) ✗ cargo check
warning: /Users/paul/Projects/celo-threshold-bls-rs/crates/dkg-cli/Cargo.toml: file found to be present in multiple build targets: /Users/paul/Projects/celo-threshold-bls-rs/crates/dkg-cli/build.rs
warning: /Users/paul/Projects/celo-threshold-bls-rs/crates/threshold-bls-ffi/Cargo.toml: unused manifest key: target.aarch64-linux-android.ar
warning: /Users/paul/Projects/celo-threshold-bls-rs/crates/threshold-bls-ffi/Cargo.toml: unused manifest key: target.aarch64-linux-android.linker
warning: unused return value of `Box::<T>::from_raw` that must be used
   --> crates/threshold-bls-ffi/src/ffi.rs:499:5
    |
499 |     Box::from_raw(token);
    |     ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_must_use)]` on by default
    = note: call `drop(from_raw(ptr))` if you intend to drop the `Box`

warning: unused return value of `Box::<T>::from_raw` that must be used
   --> crates/threshold-bls-ffi/src/ffi.rs:509:5
    |
509 |     Box::from_raw(keys);
    |     ^^^^^^^^^^^^^^^^^^^^
    |
    = note: call `drop(from_raw(ptr))` if you intend to drop the `Box`

warning: unused return value of `Box::<T>::from_raw` that must be used
   --> crates/threshold-bls-ffi/src/ffi.rs:519:5
    |
519 |     Box::from_raw(keypair);
    |     ^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: call `drop(from_raw(ptr))` if you intend to drop the `Box`

warning: unused return value of `Box::<T>::from_raw` that must be used
   --> crates/threshold-bls-ffi/src/ffi.rs:529:5
    |
529 |     Box::from_raw(private_key);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: call `drop(from_raw(ptr))` if you intend to drop the `Box`

warning: unused return value of `Box::<T>::from_raw` that must be used
   --> crates/threshold-bls-ffi/src/ffi.rs:549:5
    |
549 |     Box::from_raw(public_key);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: call `drop(from_raw(ptr))` if you intend to drop the `Box`

warning: unused return value of `Box::<T>::from_raw` that must be used
   --> crates/threshold-bls-ffi/src/ffi.rs:559:5
    |
559 |     Box::from_raw(signature);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: call `drop(from_raw(ptr))` if you intend to drop the `Box`

warning: `threshold-bls-ffi` (lib) generated 6 warnings
   Compiling dkg-cli v0.3.0 (/Users/paul/Projects/celo-threshold-bls-rs/crates/dkg-cli)
error: failed to run custom build command for `dkg-cli v0.3.0 (/Users/paul/Projects/celo-threshold-bls-rs/crates/dkg-cli)`

Caused by:
  process didn't exit successfully: `/Users/paul/Projects/celo-threshold-bls-rs/target/debug/build/dkg-cli-1567b723f449fe8b/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=../../solidity/

  --- stderr
  thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', crates/dkg-cli/build.rs:19:61
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  ```

My versions:
```
➜  celo-threshold-bls-rs git:(straka/arkworks) ✗ solc --version
solc, the solidity compiler commandline interface
Version: 0.8.17+commit.8df45f5f.Darwin.appleclang
➜  celo-threshold-bls-rs git:(straka/arkworks) ✗ rustc --version
rustc 1.64.0 (a55dd71d5 2022-09-19)
```

File: crates/dkg-cli/dkg.bin [hunk 11]
```diff
@@ -1 +1 @@
...5b9050919050565b6000602082019050818103600083015261210f81611e85565b9050919050565b6000602082019050818103600083015261212f81611eeb565b9050919050565b600060208201905061214b6000830184611f2b565b92915050565b60006040820190506121666000830185611f2b565b81810360208301526121788184611add565b90509392505050565b6000819050602082019050919050565b6000819050602082019050919050565b600081519050919050565b600081519050919050565b600081519050919050565b6000602082019050919050565b6000602082019050919050565b600082825260208201905092915050565b600082825260208201905092915050565b600082825260208201905092915050565b600082825260208201905092915050565b600082825260208201905092915050565b600061223c82612256565b9050919050565b6000819050612251826122d6565b919050565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000819050919050565b600061228b82612243565b9050919050565b60005b838110156122b0578082015181840152602081019050612295565b838111156122bf576000848401525b50505050565b6000601f19601f8301169050919050565b600381106122e057fe5b50565b6122ec81612231565b81146122f757600080fd5b50565b61230381612276565b811461230e57600080fd5b5056fea2646970667358221220c0edec4739250c6c309766472d6c9e0278b335effb542994cdc9f5f327a6466e64736f6c63430006060033
```
Comment (piersy) [inline crates/dkg-cli/dkg.bin:1]: I guess you didn't want to commit this binary file, it's probably a good idea to add it to the gitignore.

File: solidity/artifacts/DKG.sol/DKG.json [hunk 12]
```diff
@@ -0,0 +1,367 @@
+{
```
Comment (piersy) [inline solidity/artifacts/DKG.sol/DKG.json:1]: Is this addition intentional? Given that the project is intended to build this from the `.sol` file I don't see a reason to check it in. Also having this checked in opeens up the possibility that the `.sol` file can diverge from the `json`.

File: crates/dkg-cli/Cargo.toml [hunk 13]
```diff
@@ -1,15 +1,21 @@
...
 name = "dkg-cli"
-version = "0.1.0"
-authors = ["Georgios Konstantopoulos <me@gakonst.com>"]
-edition = "2018"
+version = "0.3.0"
+authors = ["Georgios Konstantopoulos <me@gakonst.com>", "Michael Straka <mstraka100@gmail.com>"]
+edition = "2022"
+build = "build.rs"
+
+[[bin]]
```
Comment (piersy) [inline crates/dkg-cli/Cargo.toml:8]: It seems I'm unable to (after changing the edition to 2021) `cargo build` this project it seems the problem is in `build.rs`

```
error: failed to run custom build command for `dkg-cli v0.3.0 (/Users/piersy/projects/celo-threshold-bls-rs/crates/dkg-cli)`

Caused by:
  process didn't exit successfully: `/Users/piersy/projects/celo-threshold-bls-rs/target/debug/build/dkg-cli-49e95d0c07e2951c/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=../../solidity/

  --- stderr
  thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', crates/dkg-cli/build.rs:19:61
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I also noticed that `Cargo.lock` was updated by the build, which feels like it should not happen, but could be a consequence of changing the edition in `dkg-cli`

I dug into the errors a bit and it seems that compilation can fail but the result is still an ok result, the compiler output needs to be introspected in greater depth to see if there were errors. When I did get the build script to run  I then faced a host more errors inside the `dkg-cli` crate.

I tried to push my bracnh up but I don't have permissions right now.

Comment (palango) [reply crates/dkg-cli/Cargo.toml:8]: I had the same, the mapping seems to require a full path in the config.

For the new errors, it seems like [this file](https://github.com/celo-org/celo-threshold-bls-rs/pull/108/files#diff-51a347bc227b1033b7aa387d6a55c3e63beae636926c9fd93e40a02183d2d210) was deleted and is missing now. Bringing it back fixes the build for me.

File: crates/dkg-cli/Cargo.toml (hunk 3 above)
Comment (mstraka100) [reply crates/dkg-cli/Cargo.toml]: Pinned ethers-rs to 0.17.0

**Review by palango — APPROVED** [review]
Looks good to me!

File: crates/dkg-cli/Dockerfile [hunk 14]
```diff
@@ -1,13 +0,0 @@
-FROM circleci/rust:1.44.0
```
Comment (piersy) [inline crates/dkg-cli/Dockerfile:1]: Why did you remove the dockerfile?

File: crates/dkg-cli/build.rs [hunk 15]
```diff
@@ -1,24 +1,40 @@
-use ethers::{contract::Abigen, utils::Solc};
```
Comment (piersy) [inline crates/dkg-cli/build.rs:1]: Why did you remove this?

File: crates/dkg-cli/blockchain_test.sh [hunk 16]
```diff
@@ -1,65 +0,0 @@
-#!/bin/bash
```
Comment (piersy) [inline crates/dkg-cli/blockchain_test.sh:1]: Why did you remove this?

File: crates/dkg-cli/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,13 @@
+FROM circleci/rust:1.44.0
```
Comment (palango) [inline crates/dkg-cli/Dockerfile:1]: This needs to be updated, we require 1.64 now

File: crates/dkg-cli/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+ENV HOME=/home/circleci
+ENV PATH=$HOME/bin:$PATH
+
+RUN cd $HOME && git clone https://github.com/celo-org/celo-threshold-bls-rs
+RUN mkdir $HOME/bin && wget -q https://github.com/ethereum/solidity/releases/download/v0.6.6/solc-static-linux -O $HOME/bin/solc && chmod u+x $HOME/bin/solc && solc --version
+RUN cd $HOME && cd celo-threshold-bls-rs/crates/dkg-cli && RUSTFLAGS="-C target-feature=-crt-static" cargo build --release
+
+FROM ubuntu:18.04
```
Comment (palango) [inline crates/dkg-cli/Dockerfile:10]: Not sure here, but I'd probably also use the latest LTS here
